Registry Authentication Spec (#192)
notaryproject/notation: specs/registry-auth.md

# Registry Authentication

> Registry access is required for pulling the manifests of the images to be verified along with their signatures as well as other advanced operations. This documentation specifies how authentication to the remote registry works and how registry credentials are stored.

Reviewer: Suggesting some rewording, feel free to modify further.

```suggestion
Registry access is required during signature generation and signature verification workflows. Based on the workflow, write access is required to push signatures, and read access is required to pull manifests, and signatures associated with a given manifest. This documentation specifies how authentication to the remote registry works and how registry credentials are stored.

```

Author: Registry access is not a MUST during signature generation and signature verification workflows, especially for local signing scenarios.

## Communication Channel

> Although it is secure to transmit artifacts with their signatures via HTTP connections, it is RECOMMENDED to transmit via HTTPS connections for confidentiality and the authenticity of the remote server.

Reviewer: Not sure why we say this, at least on public networks, HTTP is vulnerable to MITM attack where attacker can spoof and respond with valid looking descriptors and associated content.

Author: > Not sure why we say this, at least on public networks, HTTP is vulnerable to MITM attack where attacker can spoof and respond with valid looking descriptors and associated content.

This is exactly what **authenticity** means.

Reviewer: ```suggestion
Although it is secure to transmit artifacts with their signatures via HTTP connections (as a tampered artifact can be detected through signature verification), it is RECOMMENDED to transmit via HTTPS connections for confidentiality and the authenticity of the remote server.

```

Author: Updated.

> Alternatively, clients can be authenticated via mutual TLS authentication (mTLS). In other words, clients connect to servers via HTTP over mTLS by presenting client certificates. In this case, authorization can be applied later without further authentication schemes.

Reviewer: Is mTLS used/supported for registry access? If not I'd remove this content.

## Authentication Schemes

> Notation supports [basic HTTP authentication][RFC7617] scheme and token schemes including [OAuth 2.0][RFC6749] and [Docker Registry V2 Token][token]. Clients MUST connect to the remote servers via HTTPS if any authentication scheme is applied.

Reviewer: Will we enforce this if given registry URL is included in config.json`insecureRegistries` setting?

Author: Will change `MUST` to `SHOULD` to cover this.

4. The authorization grants and returns a bearer token for access back to the notation client.
5. Notation attempts to access the remote registry again with the obtained bearer token (or local credential for *basic* scheme) in the [Authorization](https://datatracker.ietf.org/doc/html/rfc7235#section-4.2) header.
6. The remote registry performs the requested operation and returns the response.

Optimization might be performed to reduced the number of requests in order to reduce the overall latency for subsequent requests.

### Basic Scheme

Notation follows [RFC 7617][RFC7617] for the *Basic* HTTP authentication scheme.

If a remote registry is known to support `Basic` authentication scheme, the attempt-challenge phase (*steps 1-4*) can be skipped, and thus Notation can access the remote registry without overhead in terms of the number of requests.

```mermaid
flowchart LR;
    client[Notation];
    reg[Registry];

    client -- Access with credentials --> reg;
       reg -- Response                --> client;
```

### Token Scheme

> Notation supports two types of token schemes, [Docker][token] and [OAuth 2.0][RFC6749] where OAuth 2.0 is preferred by default.

Reviewer 2: What does "preferred by default" mean here? Won't registries typically support one or the other?

I'm trying to understand what a client preference translates to in terms of implementation choices.

2. The remote registry returns `401 Unauthorized` with a challenge, requiring a valid token.
3. Notation requests the authorization service for a new bearer token for accessing the target resource with local credentials.
4. The authorization grants and returns a new bearer token for access back to the notation client. The client-side cache is refreshed by the newly returned token.
5. Notation attempts to access the remote registry again.
6. The remote registry performs the requested operation and returns the response.

#### Docker

In the [Docker Token Authentication][token] specification, the *step 3* is implemented using a `GET` request where users are authenticated by the authorization service via `Basic` authentication scheme. Therefore, user credentials are only accepted in the form of username and password pair.

#### OAuth 2

Notation follows the [Docker Registry v2 authentication][oauth2] specification for the [OAuth 2.0][RFC6749] framework where the *step 3* is implemented using a `POST` request. Precisely, notation supports `password` and `refresh_token` grant types.

**Note** Refresh tokens are often known as identity tokens in the context of registry authentication.

> ## Credential Store

Reviewer: We probably need a threat model section for credential store, we can add a placeholder and add details in a subsequent PR.

As local credentials may be required to access the remote registries, they need to be stored and accessed securely. To achieve maximum security, credential helpers are preferred so that credentials are stored in the system key chain with better protection. If credential helpers are not available, notation will fall back to credential files with proper access control.

### Credential Helper

> To achieve maximum compatibility with existing systems, [docker credential helpers](https://github.com/docker/docker-credential-helpers) and its [protocol](https://docs.docker.com/engine/reference/commandline/login/#credential-helper-protocol) are adopted as the credential helpers for `notation`.

Reviewer: Which [cred helper](https://github.com/docker/docker-credential-helpers#development) commands will notation use, only `get`?

Author: `get` for registry access, and `store` for `notation login` (see #119)

> To achieve maximum compatibility of existing systems, [docker credential helpers](https://github.com/docker/docker-credential-helpers) as long as its [protocol](https://docs.docker.com/engine/reference/commandline/login/#credential-helper-protocol) are adopted as the credential helpers for `notation`.

Reviewer 2: Minor nit, but I would reword this a bit.

```suggestion
To achieve maximum compatibility with existing systems, [docker credential helpers](https://github.com/docker/docker-credential-helpers) and its [protocol](https://docs.docker.com/engine/reference/commandline/login/#credential-helper-protocol) are adopted as the credential helpers for `notation`.
```

The credential store can be specified globally or per registry by setting the notation config.

```json
{
    "credHelpers": {
        "registry.wabbit-networks.io": "wabbithelper",
        "another.wabbit-networks.io": "foobar"
    },
    "credsFile": "/absolute/path/to/auth.json",
    "credsStore": "whatever"

> }

Reviewer: For my understanding, `credHelpers` can only be used if a registry is specified, and `credStrore` is the global *helper* ?

Author: Exactly.

```

> **Note** The absolute path to credential file `credsFile` is used to store extra metadata by credential helper drivers such as [docker/cli](https://github.com/docker/cli/blob/master/cli/config/credentials/native_store.go).

Reviewer: What kind of data is stored in `credsFile`?

### Credential File

The credential file is alternative credential store when credential helpers are not available. The default file path is

```
{CONFIG}/notation/auth.json
```

The credential file path can be altered by setting the `credsFile` field of the notation config.

```json
{
    "credsFile": "/absolute/path/to/auth.json"
}
```

> Since credentials are stored in plaintext, the permission of the credential file MUST be kept minimum when storing credentials. On Unix / Linux, the permission MUST be either `0600` (default) or `0400` (read-only).

Reviewer: Do we need to support local credential file, given cred helper provide a easy path to integrate both local and remote credentials?

Reviewer: Tracked in #206

> ### Credential File
> 
> The credential file is alternative credential store when credential helpers are not available. The default file path is
> 
> ```
> {CONFIG}/notation/auth.json
> ```
> 
> The credential file path can be altered by setting the `credsFile` field of the notation config.
> 
> ```json
> {
>     "credsFile": "/absolute/path/to/auth.json"
> }
> ```
> 
> Since credentials are stored in plaintext, the permission of the credential file MUST be kept minimum when storing credentials. On Unix / Linux, the permission MUST be either `0600` (default) or `0400` (read-only).

Reviewer: ```suggestion
### Credential File

TODO: Define local credential experience without credential provider support (#206).

```

Author: Updated
